File: app_pojav_zh/src/main/java/com/movtery/pojavzh/feature/ProfileLanguageSelector.java
package com.movtery.pojavzh.feature;

import static com.movtery.pojavzh.utils.ZHTools.containsDot;
import static com.movtery.pojavzh.utils.ZHTools.extractNumbers;
import static com.movtery.pojavzh.utils.ZHTools.getGameDirPath;
import static net.kdt.pojavlaunch.Tools.read;

import com.movtery.pojavzh.ui.subassembly.customprofilepath.ProfilePathHome;
import com.movtery.pojavzh.utils.ZHTools;

import net.kdt.pojavlaunch.JMinecraftVersionList;
import net.kdt.pojavlaunch.Logger;
import net.kdt.pojavlaunch.Tools;
import net.kdt.pojavlaunch.prefs.LauncherPreferences;
import net.kdt.pojavlaunch.value.launcherprofiles.MinecraftProfile;

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class ProfileLanguageSelector {
    private ProfileLanguageSelector() {
    }

    private static String getOlderLanguage(String lang) {
        StringBuilder builder = new StringBuilder(lang);
        int underscoreIndex = lang.indexOf('_');

        if (underscoreIndex != -1) {
            for (int i = underscoreIndex; i < lang.length(); i++) {
                builder.setCharAt(i, Character.toUpperCase(lang.charAt(i)));
            } //只将下划线后面的字符转换为大写
        }

        return builder.toString();
    }

    private static int getVersion(String versionId) throws NumberFormatException {
        int firstDotIndex = versionId.indexOf('.');
        int secondDotIndex = versionId.indexOf('.', firstDotIndex + 1);
        int version;

        if (firstDotIndex != -1) { // 官方版本
            if (secondDotIndex == -1) version = Integer.parseInt(versionId.substring(firstDotIndex + 1));
                else version = Integer.parseInt(versionId.substring(firstDotIndex + 1, secondDotIndex));
        } else version = 12;
        return version;
    }

    private static String getLanguage(String versionName, String rawLang) {
        if (versionName == null || rawLang == null) return null;
        String lang = rawLang;
        if (rawLang.equals("system")) lang = ZHTools.getSystemLanguage();

        JMinecraftVersionList.Version version;
        try {
            version = Tools.GLOBAL_GSON.fromJson(read(ProfilePathHome.getVersionsHome() + "/" + versionName + "/" + versionName + ".json"), JMinecraftVersionList.Version.class);
        } catch (IOException e) {
            throw new RuntimeException(e);
        }

        String versionId = version.id;

        String regex = "^\\d+[a-zA-Z]\\d+[a-zA-Z]$";
        Pattern pattern = Pattern.compile(regex);
        Matcher matcher = pattern.matcher(versionId);

        if(containsDot(versionId)) {
            try {
                int ver = getVersion(versionId);

                // 1.10 -
                if (ver < 11) {
                    return getOlderLanguage(lang);
                }

                return lang; // ? & 1.0
            } catch (NumberFormatException e) {
                return lang;
            }
        } else if (matcher.matches()) { // 快照版本 "24w09a" "16w20a"
            try {
                int[] result = extractNumbers(versionId, 2);

                if(result[0] < 16) {
                    return getOlderLanguage(lang);
                } else if (result[0] == 16 & result[1] <= 32) {
                    return getOlderLanguage(lang);
                }

                return lang;
            } catch (NumberFormatException e) {
                return lang;
            }
        }

        return lang;
    }

    public static void setToChinese(MinecraftProfile minecraftProfile) {
        File optionFile = new File((getGameDirPath(minecraftProfile.gameDir)), "options.txt");
        ArrayList<String> options = new ArrayList<>();
        boolean foundMatch = false;
        String language = getLanguage(minecraftProfile.lastVersionId, LauncherPreferences.PREF_GAME_LANGUAGE);

        try (BufferedReader optionFileReader = new BufferedReader(new InputStreamReader(new FileInputStream(optionFile), StandardCharsets.UTF_8))) {
            String line;
            while ((line = optionFileReader.readLine()) != null) {
                //使用正则表达式匹配“lang: xxx”格式
                Pattern pattern = Pattern.compile("lang:(\\S+)");
                Matcher matcher = pattern.matcher(line);

                if (matcher.find()) {
                    foundMatch = true;
                }

                options.add(line);
            }
        } catch (IOException e) {
            Logger.appendToLog(e.toString());
            return;
        }

        if (!foundMatch) {
            options.add("lang:" + language);
            try (BufferedWriter optionFileWriter = new BufferedWriter(new OutputStreamWriter(new FileOutputStream(optionFile), StandardCharsets.UTF_8))) {
                for (String option : options) {
                    optionFileWriter.write(option);
                    optionFileWriter.newLine();
                }
            } catch (IOException e) {
                Logger.appendToLog(e.toString());
            }
        }
    }
}

File: app_pojav_zh/src/main/java/com/movtery/pojavzh/feature/ProfileLanguageSelector.kt
package com.movtery.pojavzh.feature

import com.movtery.pojavzh.feature.customprofilepath.ProfilePathHome.Companion.versionsHome
import com.movtery.pojavzh.feature.log.Logging
import com.movtery.pojavzh.feature.mod.modloader.NeoForgeUtils
import com.movtery.pojavzh.utils.MCVersionRegex
import com.movtery.pojavzh.utils.ZHTools
import net.kdt.pojavlaunch.JMinecraftVersionList
import net.kdt.pojavlaunch.Tools
import net.kdt.pojavlaunch.prefs.LauncherPreferences
import net.kdt.pojavlaunch.utils.MCOptionUtils
import net.kdt.pojavlaunch.value.launcherprofiles.MinecraftProfile
import org.jackhuang.hmcl.util.versioning.VersionNumber
import org.jackhuang.hmcl.util.versioning.VersionRange


class ProfileLanguageSelector {
    companion object {
        private fun getOlderLanguage(lang: String): String {
            val underscoreIndex = lang.indexOf('_')
            return if (underscoreIndex != -1) {
                //只将下划线后面的字符转换为大写
                val builder = StringBuilder(lang.substring(0, underscoreIndex + 1))
                builder.append(lang.substring(underscoreIndex + 1).uppercase())
                builder.toString()
            } else lang
        }

        private fun getLanguage(versionName: String, rawLang: String): String {
            val lang = if (rawLang == "system") ZHTools.getSystemLanguage() else rawLang

            val version: String = runCatching {
                val versionObject = Tools.GLOBAL_GSON.fromJson(
                    Tools.read("$versionsHome/$versionName/$versionName.json"),
                    JMinecraftVersionList.Version::class.java
                )
                versionObject.id
            }.getOrElse { e ->
                Logging.e("ProfileLanguageSelector", "Failed to read version data : \n ${Tools.printToString(e)}")
                "1.11"
            }

            val versionId = VersionNumber.asVersion(version).canonical
            Logging.i("ProfileLanguageSelector", "Version Id : $versionId")

            return when {
                versionId.contains('-') -> otherVersion(versionId, lang) // Mod加载器或者其他
                versionId.contains('.') -> {
                    if (isOlderVersionRelease(versionId)) getOlderLanguage(lang) // 1.10 -
                    else lang
                }
                MCVersionRegex.SNAPSHOT_REGEX.matcher(versionId).matches() -> { // 快照版本 "24w09a" "16w20a"
                    if (isOlderVersionSnapshot(versionId)) getOlderLanguage(lang)
                    else lang
                }
                else -> lang
            }
        }

        private fun otherVersion(versionId: String, lang: String): String {
            // 定义关键词，以及相应的版本名称处理逻辑
            val suffixes = mapOf(
                // "1.20.4-OptiFine_HD_U_I7_pre3"   -> "1.20.4"
                "-OptiFine" to { id: String -> id.substringBefore('-') },
                // "1.20.2-forge-48.1.0"            -> "1.20.2"
                "-forge-" to { id: String -> id.substringBefore('-') },
                // "neoforge-21.1.8"                -> "1.21.1"
                "neoforge-" to { id: String -> NeoForgeUtils.formatGameVersion(id.removePrefix("neoforge-")) },
                // "fabric-loader-0.15.7-1.20.4"    -> "1.20.4"
                "fabric-loader-" to { id: String -> id.substringAfterLast('-') },
                // "quilt-loader-0.23.1-1.20.4"     -> "1.20.4"
                "quilt-loader-" to { id: String -> id.substringAfterLast('-') }
            )

            val versionName = suffixes.entries
                .firstOrNull { versionId.contains(it.key, true) }
                ?.value?.invoke(versionId)
                ?: versionId //其他情况，命名规则千变万化，故不做检测

            Logging.i("ProfileLanguageSelector", "Version Name: $versionName")

            return if (isOlderVersionRelease(versionName)) getOlderLanguage(lang) else lang
        }

        private fun isOlderVersionRelease(versionName: String): Boolean {
            return VersionRange.atMost(VersionNumber.asVersion("1.10.2")).contains(VersionNumber.asVersion(versionName))
        }

        private fun isOlderVersionSnapshot(versionName: String): Boolean {
            return VersionRange.atMost(VersionNumber.asVersion("16w32a")).contains(VersionNumber.asVersion(versionName))
        }

        @JvmStatic
        fun setGameLanguage(minecraftProfile: MinecraftProfile, overridden: Boolean) {
            if (MCOptionUtils.containsKey("lang") && !overridden) return
            val language = getLanguage(minecraftProfile.lastVersionId, LauncherPreferences.PREF_GAME_LANGUAGE)
            Logging.i("ProfileLanguageSelector", "The game language has been set to: $language")
            MCOptionUtils.set("lang", language)
        }
    }
}
